cpp/include/cudf/detail/gather.cuh

```diff
@@ -46,56 +46,104 @@ struct bounds_checker {
-
-      while (destination_row_base < destination_table_num_rows) {
-        size_type destination_row = destination_row_base + threadIdx.x;
+    while (bit_idx_offset < mask_num_bits) {
+      auto const bit_idx = bit_idx_offset + threadIdx.x;
+      auto const bit_active = bit_idx < mask_num_bits;
+      auto const bit_is_valid = bit_active && get_validity(mask_idx, bit_idx);
+      auto const warp_validity = __ballot_sync(0xffffffff, bit_is_valid);
+      auto const validity_idx = word_index(bit_idx);
 
-        const bool thread_active = destination_row < destination_table_num_rows;
-        size_type source_row =
-          thread_active ? gather_map[destination_row] : 0;
+      if (threadIdx.x % warp_size == 0) {
+        mask[validity_idx] = warp_validity;
+      }
 
-        bool bit_is_valid;
-        if (ignore_out_of_bounds && (source_row < 0 || source_row >= source_col.size())) {
-          bit_is_valid = thread_active && bit_is_set(masks[i], destination_row);
-        } else {
-          bit_is_valid = source_col.is_valid(source_row);
-        }
+      validity_count_aggregate += single_lane_block_popc_reduce(warp_validity);
+      bit_idx_offset += blockDim.x * gridDim.x;
+    }
 
-        // Use ballot to find all valid bits in this warp and create the output
-        // bitmask element
-        const uint32_t valid_warp =
-          __ballot_sync(0xffffffff, thread_active && bit_is_valid);
 
-        const size_type valid_index = word_index(destination_row);
+    if (threadIdx.x == 0) {
+      atomicAdd(validity_counts + mask_idx, validity_count_aggregate);
+    }
+  }
+}
 
-        // Only one thread writes output
-        if (0 == threadIdx.x % warp_size and thread_active) {
-          masks[i][valid_index] = valid_warp;
-        }
-        valid_count_accumulate += single_lane_block_popc_reduce(valid_warp);
-        destination_row_base += blockDim.x * gridDim.x;
-      }
-      if (threadIdx.x == 0) {
-        atomicAdd(valid_counts + i, valid_count_accumulate);
+template <bool ignore_out_of_bounds, typename MapIterator>
+struct gather_bitmask_functor {
+  table_device_view input;
+  bitmask_type** masks;
+  MapIterator gather_map;
+
+  __device__ bool operator()(size_type mask_idx, size_type bit_idx) {
+    auto row_idx = gather_map[bit_idx];
+    auto col = input.column(mask_idx);
+
+    if (ignore_out_of_bounds) {
+      if (row_idx < 0 || row_idx >= col.size()) {
+        return bit_is_set(masks[mask_idx], bit_idx);
       }
     }
+
+    return col.is_valid(row_idx);
   }
+};
+
+template <typename ValiditySelector>
+__global__ void select_bitmask_kernel(ValiditySelector get_validity,
+                                      bitmask_type* masks[],
+                                      size_type mask_count,
+                                      size_type mask_num_bits,
+                                      size_type* validity_counts) {
+  select_bitmask(get_validity, masks, mask_count, mask_num_bits, validity_counts);
```

Comment:
It's a bit unusual to have a kernel that just calls a device function. Can the body of the function just be moved into the kernel? 

I suspect this was done to preserve the existing `gather_bitmask_kernel`, but that can just be factored out to instead call the new kernel.

Reply:
Sounds good to me! I wasn't sure how much we wanted to change in a single PR. Right now `scatter` and `gather` are the only functions using the gather kernel. I'll remove `__global__ void gather_bitmask_kernel` and move the Selector initialization up a level.